### DISCUSSION TITLE
Filter hooks broken?

### DISCUSSION THREAD

**Discussion by sebj54** [directus/directus#10486] [Archive]
I am actually trying to build a filter hook on `items.read`. The point is to force filters on the payload, given on user's permissions.

The docs makes think it is possible, and I have been doing this several times in other projects (v9 but also v8). Here is what the docs say about filter hooks:

> The callback function itself receives three parameters:
> 
> - The modifiable payload
> - An event-specific meta object
> - A context object

But this first parameter does not look to be the payload, but rather the records returned in the API response.

### To Reproduce

Create a custom hook (`extensions/hooks/test/index.js`):

```js
module.exports = function registerHook({ filter }, { services, exceptions }) {
    filter('items.read', async (payload, { collection }, { schema, accountability }) => {
        console.log('[TCL] ~ file: test/index.js ~ line 3 ~ filter ~ payload', payload)

        return payload
    })
}
```

Run Directus, then go to your browser or Postman, make a GET request on any collection. Check your console, there should be items logged as the `payload` object.

### Errors Shown

There is no error shown, but see the API call and the payload logged:

![image](https://user-images.githubusercontent.com/5880435/145859779-37f5fe31-02da-4bc8-b03b-8f263f9add8c.png)

![image](https://user-images.githubusercontent.com/5880435/145860125-b3219ad5-0ff8-4146-92fd-2710593bd45e.png)

**Comment by rijkvanzanten** [comment]
Heya! This is actually working as expected. The items.read hook is meant to alter the payload that's sent to the user on reads, not to alter the query that was submitted by the user.

> and I have been doing this several times in other projects (v9 but also v8)

Curious to learn more about this! How were you doing it in v9 before? The `read` filter was only introduced very recently, and hasn't changed in how it operates afaik.

**Comment by sebj54** [comment]
My bad, you're right! I've only modified payloads on create or update hooks before (and also it was still the alpha, before `filter` and `action` hooks).

So it is not possible to modify payloads on a read action? It seems very counter-intuitive, especially when the [docs](https://docs.directus.io/extensions/hooks/#filter) state:

> Filter hooks act on the event's payload before the event is fired. 

So if you modify the paylaod returned to the user, it actually happens after the event is fired and you don't really cancel the event. 

I see that as a problem because if you have to filter the returned values (when listing a collection), you may break pagination, especially on large datasets.

What are the use cases then? I only see filtering the props of the objects returned or maybe add some computed values. 

<hr>

Is there any way to alter the query submitted by the user? I've done it before to [add the raw body in a custom endpoint](https://github.com/directus/directus/discussions/8633#discussioncomment-1454358) with `middlewares.init.before` but it may have changed with the new way of registering hooks.

Maybe there is room for a new `query` hook? 😇

**Reply by rijkvanzanten** [reply]
> What are the use cases then? I only see filtering the props of the objects returned or maybe add some computed values.

This is exactly it! Transforming the data points before returning them to the user. 

> Maybe there is room for a new query hook? 😇

This sounds like the best way to go to me!

**Reply by sebj54** [reply]
Ok great! Do you plan to develop it on your side or do you want a PR? I will probably have time to implement it by the end of week. As I am not very familiar with the directus codebase, feel free to give me hints (like the files I will to have to edit).

**Reply by rijkvanzanten** [reply]
A PR or sponsoring the implementation would be the fastest way to get the functionality added. We can't guarantee any timing on open feature requests otherwise.

This is where it'd have to be added:

https://github.com/directus/directus/blob/5eafb62fea629d0ee8a92b852113e4494a48f1e7/api/src/services/items.ts#L258-L320

The actual code to implement it should be nearly identical to the statement on 291

**Reply by sebj54** [reply]
Thank you very much for the advice 🙏 

I would love to sponsor (and the whole team really deserves it!), but sadly my project has already exceeded the time budget 😞 I will mark this discussion as answered for now. See you soon on a PR! :)
